.azure-pipelines/advanced.yml

```diff
@@ -11,8 +11,15 @@ schedules:
       - master
     always: true
 
-jobs:
-  # Any addition here should be reflected in the advanced-test and release pipelines.
-  # It is advised to declare all jobs here as templates to improve maintainability.
-  - template: templates/tests-suite.yml
-  - template: templates/installer-tests.yml
+stages:
+  - stage: Test_Package
+    jobs:
+      # Any addition here should be reflected in the advanced-test and release pipelines.
+      # It is advised to declare all jobs here as templates to improve maintainability.
+      - template: templates/jobs/standard-tests-jobs.yml
+      - template: templates/jobs/extended-tests-jobs.yml
+      - template: templates/jobs/packaging-jobs.yml
+  - stage: Deploy
+    jobs:
+      - template: templates/jobs/deployment-jobs.yml
```

Comment:
Indeed. My bad ...

Reply:
Update, in fact there is a discussion to do here.

Initially, the `advanced` pipeline on Azure serves two independant purpose:
1) nightly builds of master
2) pre-check release for the release branches

It was ok for Windows specifically because there were not deployment to do in the two situations.

On Linux however, the deployment of the snaps to edge must occur for 1), but not for 2).

There are several possibilities to fix the problem here:
a. we can split the pipeline in two, `nightly` for 1), `advanced` for 2)
b. we can consider that 2) is not needed, because the complete tests are already done for release tags, and so leave only a `nightly` for 1)
c. we can move the trigger for release branches in the `advanced-test` pipeline, that serves the same purpose (minus the Mattermost notification, that could either be dropped or added to `advanced-test` and disabled conditionnally for test branches).

My preference goes for b., if it does not have any impact to the actual release logic, since it removes complexity instead of adding anything. As an alternative, a. is the second best since stages are highly reusable with minimal added complexity.

What do you think?

Reply:
I don't think we should do b due to our current release process.

In order for us to push a correct tag for a release right now, we need to have already done the majority of the 1+ hour manual release process. This is largely because the tag includes an updated version of certbot-auto for the release which is signed and includes the hashes of our built Python packages for the release.

Before doing this manual process, I think we should run our full test suite on the closest thing to the code we want to release as we can. Not doing this can result in problems that would have been caught before manually building the release being caught afterwards and we have to repeat the process. I'm not OK with adding this potential headache for Erica and I to the release process.

With our current release script, the closest thing to the code we want to release for point releases is are *.x branches. We could trigger the builds manually before a release, but I think we are all trying to automate things more and not less so I don't think this is a good approach either.

For the other options, I don't have that strong of a preference, but I will say that in general:

1. I would like to keep Mattermost notifications for *.x branch failures because in most cases it's signalling test failures on the code we're preparing to release. I'd like Azure to make noise about that.
2. Between splitting pipelines and adding conditions, there is of course a balance to be struck, but when we're talking about adding a single condition, I personally prefer that over splitting pipelines. I think that results in fewer config lines and duplication and I think it is easier to read and understand our test setup rather than having it all spread over multiple files.

I'll also add option d which would be to skip the deploy stage in `advanced.yml` for `*.x` branches (or if it's not triggered by cron). This may mean undoing the change to make a reusable deploy stage.

Reply:
Ah yes, I forget each time that release tags are actually the end of the release process, not its beginning. Sorry for the noise.

Your option d is quite decent also, and in fact as for c it is not necessary to remove the abstraction of the deploy stage/mattermost notifications stage. Indeed, you can override any parameter of a template used in a pipeline, so it is possible to add a `condition` which is not set in the stage template itself.

However between c and d, I think c is better for two reasons. First it allows to have pipelines with a coherent purpose (night builds for `advanced`, that could be renamed `nightly`, and advanced tests for specific branches with `advanced-tests`). Second, and more importantly, option d is more error prone on a critical part which is the deploy phase: indeed in this case it is a condition that decide if snaps will be pushed or not. If we change the outcome of the tested condition, directly or indirectly, it could end with accidental deployment of testing branches. Oups.

Reply:
Given your preference to not have too much pipelines, and given what I said in the previous post, I propose to use option c, with a conditional trigger of mattermost notification for `*.x` branch.

I also propose to rename the `advanced.yml` file into `nightly.yml`, and reflect these changes in Azure Pipeline configuration for certbot.

Reply:
SGTM!

Reply:
I did that.